cdc_admin/cdc_admin_handler.cpp

```diff
@@ -0,0 +1,253 @@
+    e.errorCode = ::cdc_admin::CDCAdminErrorCode::INVALID_UPSTREAM;
+    e.message = folly::stringPrintf("Invalid ip:port %s:%d", ip.c_str(), port);
+    callback->release()->exceptionInThread(std::move(e));
+    return false;
+  }
+}
+
+}  // anonymous namespace
+
+namespace cdc_admin {
+
+CdcAdminHandler::CdcAdminHandler(
+    std::unique_ptr<CDCApplicationDBManager<CDCApplicationDB, replicator::DbWrapper>> db_manager)
+    : db_admin_lock_(), db_manager_(std::move(db_manager)) {}
+
+CdcAdminHandler::~CdcAdminHandler() {}
+
+std::shared_ptr<CDCApplicationDB> CdcAdminHandler::getDB(const std::string& db_name,
+                                                         CDCAdminException* ex) {
+  std::string err_msg;
+  auto db = db_manager_->getDB(db_name, &err_msg);
+  if (db == nullptr && ex) {
+    ex->errorCode = CDCAdminErrorCode::NOT_FOUND;
+    ex->message = std::move(err_msg);
+  }
+
+  return db;
+}
+
+std::unique_ptr<replicator::DbWrapper> CdcAdminHandler::removeDB(const std::string& db_name,
+                                                                 CDCAdminException* ex) {
+  std::string err_msg;
+  auto db = db_manager_->removeDB(db_name, &err_msg);
+  if (db == nullptr && ex) {
+    ex->errorCode = CDCAdminErrorCode::NOT_FOUND;
+    ex->message = std::move(err_msg);
+  }
+
+  return db;
+}
+
+
+void CdcAdminHandler::async_tm_addObserver(
+    std::unique_ptr<apache::thrift::HandlerCallback<std::unique_ptr<AddObserverResponse>>> callback,
+    std::unique_ptr<AddObserverRequest> request) {
+  db_admin_lock_.Lock(request->db_name);
+  SCOPE_EXIT { db_admin_lock_.Unlock(request->db_name); };
+
+  CDCAdminException e;
+  auto db = getDB(request->db_name, &e);
+  if (db) {
+    e.errorCode = CDCAdminErrorCode::ALREADY_EXIST;
+    e.message = "Db already exists";
+    callback.release()->exceptionInThread(std::move(e));
+    return;
+  }
+
+  // Get the upstream for the db to be added
+  auto upstream_addr = std::make_unique<folly::SocketAddress>();
+  if (!SetAddressOrException(
+          request->upstream_ip, FLAGS_rocksdb_replicator_port, upstream_addr.get(), &callback)) {
+    return;
+  }
+
+  auto segment = common::DbNameToSegment(request->db_name);
+  auto db_path = FLAGS_rocksdb_dir + request->db_name;
+
+  // add the db to db_manager
+  std::string err_msg;
+  replicator::DBRole role = replicator::DBRole::SLAVE;
+  if (request->__isset.db_role) {
+    if (request->db_role == "LEADER") {
+      role = replicator::DBRole::SLAVE;
+    } else if (request->db_role == "STANDBY") {
+      role = replicator::DBRole::NOOP;
+    } else {
+      e.errorCode = CDCAdminErrorCode::INVALID_ROLE;
+      e.message = std::move(request->db_role);
+      callback.release()->exceptionInThread(std::move(e));
+      return;
+    }
+  }
+
+  // TODO(indy): Read sequence # from kafka
+  std::unique_ptr<replicator::TestDBProxy> db_wrapper =
+      std::make_unique<replicator::TestDBProxy>(request->db_name, 0);
```

Comment:
TestDBProxy?

Reply:
Added a TODO to replace this, this is just for testing